### PR TITLE
Mentorship profiles

### PR DESCRIPTION
**Pull request by Andresgarciasch**
https://github.com/4GeeksAcademy/About-4Geeks-Academy/issues/3069

### CODE REVIEW

File: breathecode/mentorship/tests/urls/tests_academy_mentor.py [hunk 1]
```diff
@@ -674,8 +682,47 @@ def test__post__creating_a_element(self):
...
+                                   model.mentorship_service,
+                                   model.user,
+                                   data={
+                                       'id': 1,
+                                       'slug': 'mirai-nikki',
+                                       'email': model.user.email
+                                   })
+
+        self.assertEqual(json, expected)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (jefer94) [inline breathecode/mentorship/tests/urls/tests_academy_mentor.py:728]: Add test to the database something like this
```py
        self.assertEqual(self.bc.database.list_of('mentorship.MentorProfile'), [
            mentor_profile_columns({
                'id': 1,
                'name': 'Mirai Nikki',
                'slug': 'mirai-nikki',
                'bio': None,
                'user_id': 1,
                'academy_id': 1,
                'price_per_hour': 20.0,
                'email': email,
            }),
        ])
```

File: breathecode/authenticate/views.py [hunk 2]
```diff
@@ -271,6 +271,28 @@ def post(self, request, academy_id=None):
     @capable_of('crud_member')
     def put(self, request, academy_id=None, user_id_or_email=None):
 
+        mentor = MentorProfile.objects.filter(id=mentor_id, services__academy__id=academy_id).first()
```
Comment (jefer94) [inline breathecode/authenticate/views.py:274]: A staff member can not have MentorProfile, remove this validation, this is not needed

File: breathecode/authenticate/serializers.py [hunk 3]
```diff
@@ -463,6 +463,51 @@ def validate(self, data):
                     f'This user is already a member of this academy as {str(already.role)}',
                     slug='already-exists')
 
+        profile_academy = ProfileAcademy.objects.filter(user__id=data['user'].id,
```
Comment (jefer94) [inline breathecode/authenticate/serializers.py:466]:
```py
profile_academy = ProfileAcademy.objects.filter(user__id=data['user'].id, \
    academy__id=data['academy'].id, first_name__is_null=False, \
    last_name__is_null=False).exclude(first_name='', last_name='')
```

**Review by jefer94 — CHANGES_REQUESTED** [review]

**Comment by jefer94** [comment]
The tests is failing, the coverage will be saw after fix those tests

**Comment by jefer94** [comment]
tests fail

File: breathecode/registry/actions.py [hunk 4]
```diff
@@ -461,7 +461,7 @@ def get_thumbnail_url(self) -> tuple[str, bool]:
             # register click
             media.hits += 1
             media.save()
-
+            print('11111111111111111')
```
Comment (jefer94) [inline breathecode/registry/actions.py:464]: remove prints

File: breathecode/mentorship/tests/urls/tests_academy_mentor_id.py [hunk 5]
```diff
@@ -519,14 +551,20 @@ def test__post__with_one_mentor_profile__changing_to_a_success_status(self):
     def test__post__with_one_mentor_profile__changing_to_a_success_status__raise_a_exception(self):
         statuses = ['INVITED', 'ACTIVE', 'UNLISTED', 'INNACTIVE']
         valid_statuses = ['ACTIVE', 'UNLISTED']
+        profile_academy = {
+            'first_name': self.bc.fake.name(),
+            'last_name': self.bc.fake.name(),
+            'phone': self.bc.fake.phone_number(),
```
Comment (jefer94) [inline breathecode/mentorship/tests/urls/tests_academy_mentor_id.py:557]: Add a case when the mentor is being changed and the profile academy doesn't have its property set

File: breathecode/mentorship/views.py [hunk 6]
```diff
@@ -751,10 +752,30 @@ def put(self, request, mentor_id=None, academy_id=None):
...
+                                      code=400,
+                                      slug='without last-name')
+
+        if user.email is None:
+            raise ValidationException('This mentor does not have an email address',
+                                      code=400,
+                                      slug='without email')
+
+        if user.phone is None:
+            raise ValidationException('This mentor does not have a phone', code=400, slug='without phone')
```
Comment (jefer94) [inline breathecode/mentorship/views.py:778]: replace spaces by hyphen (`-`) in the slug

File: breathecode/mentorship/views.py [hunk 7]
```diff
@@ -751,10 +752,30 @@ def put(self, request, mentor_id=None, academy_id=None):
...
                                       slug='not-found')
+        user = ProfileAcademy.objects.filter(user__id=mentor.user.id, academy__id=academy_id).first()
+
+        if user.first_name is None:
+            raise ValidationException('This mentor does not have a first name',
+                                      code=400,
+                                      slug='without first-name')
+
+        if user.last_name is None:
+            raise ValidationException('This mentor does not have a last name',
```
Comment (jefer94) [inline breathecode/mentorship/views.py:768]: Add translations https://github.com/breatheco-de/apiv2/blob/0835b4b1fa060e61d95f6fc5cf72432817be9fa9/breathecode/payments/views.py#L50

File: breathecode/mentorship/tests/urls/tests_academy_mentor.py [hunk 8]
```diff
@@ -748,8 +756,59 @@ def test__post__creating_a_element(self):
...
 
         mentor_profile = self.bc.database.get('mentorship.MentorProfile', 1, dict=False)
         self.bc.check.queryset_with_pks(mentor_profile.services.all(), [1])
+
+    @patch('django.utils.timezone.now', MagicMock(return_value=UTC_NOW))
+    def test__post__creating_a_element_without_email(self):
+        profile_academy = {
+            'first_name': self.bc.fake.name(),
+            'last_name': self.bc.fake.name(),
+        }
```
Comment (jefer94) [inline breathecode/mentorship/tests/urls/tests_academy_mentor.py:771]: Add a case when the mentor is being changed and the profile academy doesn't have its property set

File: breathecode/authenticate/serializers.py [hunk 9]
```diff
@@ -463,6 +465,44 @@ def validate(self, data):
...
+        if not data['last_name'] and profile_academy:
+
+            data['last_name'] = profile_academy.last_name
+
+        if not data['last_name'] and user:
+
+            data['last_name'] = user.last_name
+
+        if not data['last_name']:
+            raise ValidationException('Unable to find last name on this user', code=400)
```
Comment (jefer94) [inline breathecode/authenticate/serializers.py:504]: Add slug and translations

File: breathecode/authenticate/serializers.py [hunk 10]
```diff
@@ -779,10 +819,38 @@ class Meta:
...
+        if not data['last_name'] and profile_academy:
+
+            data['last_name'] = profile_academy.last_name
+
+        if not data['last_name']:
+
+            data['last_name'] = User.last_name
+
+        if not data['last_name']:
+            raise ValidationException('Unable to find last name on this user', code=400)
```
Comment (jefer94) [inline breathecode/authenticate/serializers.py:852]: Add slug and translations

File: breathecode/authenticate/tests/urls/tests_academy_id_member_id.py [hunk 11]
```diff
@@ -148,12 +148,17 @@ def test_academy_id_member_id__without_capabilities(self):
         self.assertEqual(self.all_cohort_time_slot_dict(), [])
 
     def test_academy_id_member_id__with_auth(self):
+        profile_academy = {
+            'first_name': self.bc.fake.first_name(),
+            'last_name': self.bc.fake.last_name(),
+            'email': self.bc.fake.email(),
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_id_member_id.py:154]: Add a case when the ProfileAcademy is not being saved with the requires fields

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 12]
```diff
@@ -519,12 +519,20 @@ def test_academy_member_id__put__without_capability(self):
     @patch('os.getenv', MagicMock(return_value='https://dotdotdotdotdot.dot'))
     def test_academy_member_id__put__passing_email(self):
         """Test /academy/:id/member/:id"""
+
+        profile_academy = {
+            'first_name': self.bc.fake.first_name(),
+            'last_name': self.bc.fake.last_name(),
+            'email': self.bc.fake.email()
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:526]: Add a case when the ProfileAcademy is not being saved with the requires fields

File: breathecode/authenticate/tests/urls/tests_academy_student_id.py [hunk 13]
```diff
@@ -482,12 +482,17 @@ def test_academy_student_id__put__user_exists_but_without_profile_academy(self):
     @patch('os.getenv', MagicMock(return_value='https://dotdotdotdotdot.dot'))
     def test_academy_student_id__put__with_data(self):
         """Test /academy/:id/member/:id"""
+        profile_academy = {
+            'first_name': self.bc.fake.first_name(),
+            'last_name': self.bc.fake.last_name(),
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_student_id.py:487]: Add a case when the ProfileAcademy is not being saved with the requires fields

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/authenticate/tests/urls/tests_academy_student_id.py (hunk 13 above)
Comment (Andresgarciasch) [reply breathecode/authenticate/tests/urls/tests_academy_student_id.py:487]: This comment has a slight problem, which is that at the moment of execution, it tries to pass through a PUT that is validating only 'crud_member' kinda requests but this test is trying to pass 'crud_student' so I changed it to 'crud_member' anyways even knowing that something may be already testing it in another folder

Comment (jefer94) [reply breathecode/authenticate/tests/urls/tests_academy_student_id.py:487]: `crud_student` and `crud_member` are capabilities, so you need add to the test `self.bc.database.create(capability='capability', profile_academy=1, role=1)`

File: breathecode/mentorship/views.py [hunk 14]
```diff
@@ -747,14 +751,54 @@ def post(self, request, academy_id=None):
...
+                            slug='without-email'),
+                code=400,
+            )
+
+        if user.phone is None:
+            raise ValidationException(
+                translation(lang,
+                            en='This mentor does not have a phone',
+                            es='Este mentor no tiene numero de telefono',
+                            slug='without-phone'),
```
Comment (jefer94) [inline breathecode/mentorship/views.py:799]: This was not cover

File: breathecode/mentorship/views.py [hunk 15]
```diff
@@ -747,14 +751,54 @@ def post(self, request, academy_id=None):
...
+                            slug='without-last-name'),
+                code=400,
+            )
+
+        if user.email is None:
+            raise ValidationException(
+                translation(lang,
+                            en='This mentor does not have an email address',
+                            es='Este mentor no tiene correo electrónico',
+                            slug='without-email'),
```
Comment (jefer94) [inline breathecode/mentorship/views.py:790]: This was not cover

File: breathecode/mentorship/views.py [hunk 16]
```diff
@@ -747,14 +751,54 @@ def post(self, request, academy_id=None):
...
+                            slug='without-first-name'),
+                code=400,
+            )
+
+        if user.last_name is None:
+            raise ValidationException(
+                translation(lang,
+                            en='This mentor does not have a last name',
+                            es='Este mentor no tiene apellido',
+                            slug='without-last-name'),
```
Comment (jefer94) [inline breathecode/mentorship/views.py:781]: This was not cover

File: breathecode/mentorship/serializers.py [hunk 17]
```diff
@@ -492,6 +529,38 @@ def validate(self, data):
...
+        if not data['email'] and profile_academy:
+
+            data['email'] = profile_academy.email
+
+        if not data['email']:
+
+            data['email'] = data['user'].email
+
+        if not data['email']:
+            raise ValidationException('Unable to find email on this user', code=400)
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:562]: Add translations here

Comment (jefer94) [reply breathecode/mentorship/serializers.py:562]: Cover this case

File: breathecode/mentorship/serializers.py [hunk 18]
```diff
@@ -492,6 +529,38 @@ def validate(self, data):
...
+        if not data['name'] and profile_academy:
+
+            data['name'] = profile_academy.first_name + ' ' + profile_academy.last_name
+
+        if not data['name']:
+
+            data['name'] = user.first_name + ' ' + user.last_name
+        data['name'] = data['name'].strip()
+        if not data['name']:
+            raise ValidationException('Unable to find name on this user', code=400)
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:548]: Cover this case

Comment (jefer94) [reply breathecode/mentorship/serializers.py:548]: Add translations here

File: breathecode/mentorship/serializers.py [hunk 19]
```diff
@@ -471,6 +472,42 @@ class Meta:
...
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:493]: Add translations

Comment (jefer94) [reply breathecode/mentorship/serializers.py:493]: Cover these cases

File: breathecode/authenticate/views.py [hunk 20]
```diff
@@ -271,6 +271,26 @@ def post(self, request, academy_id=None):
...
+
+        if user.first_name is None:
+            raise ValidationException('This mentor does not have a first name',
+                                      code=404,
+                                      slug='first-name-not-found')
+
+        if user.last_name is None:
+            raise ValidationException('This mentor does not have a last name',
+                                      code=404,
+                                      slug='last-name-not-found')
```
Comment (jefer94) [inline breathecode/authenticate/views.py:284]: Cover these cases

File: breathecode/authenticate/serializers.py [hunk 21]
```diff
@@ -779,10 +828,46 @@ class Meta:
...
+
+        if not data['last_name']:
+
+            data['last_name'] = User.last_name
+
+        if not data['last_name']:
+            raise ValidationException(translation(lang,
+                                                  en='Unable to find last name on this user',
+                                                  es='Imposible encontrar el apellido en este usuario',
+                                                  slug='last-name-not-found'),
```
Comment (jefer94) [inline breathecode/authenticate/serializers.py:868]: Cover these cases

File: breathecode/authenticate/serializers.py [hunk 22]
```diff
@@ -463,6 +466,52 @@ def validate(self, data):
...
+
+        if not data['last_name'] and user:
+
+            data['last_name'] = user.last_name
+
+        if not data['last_name']:
+            raise ValidationException(translation(lang,
+                                                  en='Unable to find last name on this user',
+                                                  es='Imposible encontrar el apellido en este usuario',
+                                                  slug='last-name-not-found'),
```
Comment (jefer94) [inline breathecode/authenticate/serializers.py:512]: Cover these cases

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 23]
```diff
@@ -542,39 +626,189 @@ def test_academy_member_id__put__passing_email(self):
...
+                                        capability='crud_member',
+                                        profile_academy=profile_academy)
+
+        url = reverse_lazy('authenticate:academy_member_id', kwargs={'user_id_or_email': '2'})
+
+        data = {'role': role, 'first_name': self.bc.fake.first_name(), 'email': self.bc.fake.email()}
+        response = self.client.put(url, data, format='json')
+
+        json = response.json()
+        expected = {'detail': 'last-name-not-found', 'status_code': 400}
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:762]: This appear 3 times, you cover one of 3 apparitions of this exception, it is in a view and 2 serializers and it cover just one of 3 apparition

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 24]
```diff
@@ -542,39 +626,189 @@ def test_academy_member_id__put__passing_email(self):
...
+            role=role,
+            capability='crud_member',
+            profile_academy=1,
+        )
+        url = reverse_lazy('authenticate:academy_member_id', kwargs={'user_id_or_email': 'dude@dude.dude'})
 
         response = self.client.put(url)
         json = response.json()
-        expected = {'role': ['This field is required.']}
+        expected = {'detail': 'first-name-not-found', 'status_code': 404}
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:644]: 400

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 25]
```diff
@@ -542,39 +626,189 @@ def test_academy_member_id__put__passing_email(self):
...
+            model = self.bc.database.create(authenticate=True,
+                                            capability='crud_member',
+                                            role='role',
+                                            profile_academy=profile_academy)
+
+            url = reverse_lazy('authenticate:academy_member_id', kwargs={'user_id_or_email': n})
+            response = self.client.put(url)
+
+            json = response.json()
+            expected = {'detail': 'first-name-not-found', 'status_code': 404}
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:668]: 400

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 26]
```diff
@@ -542,39 +626,189 @@ def test_academy_member_id__put__passing_email(self):
...
+            model = self.bc.database.create(authenticate=True,
+                                            capability='crud_member',
+                                            role='role',
+                                            profile_academy=profile_academy)
+
+            url = reverse_lazy('authenticate:academy_member_id', kwargs={'user_id_or_email': n})
+            response = self.client.put(url)
+
+            json = response.json()
+            expected = {'detail': 'last-name-not-found', 'status_code': 404}
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:689]: 400

File: breathecode/authenticate/tests/urls/tests_academy_member_id.py [hunk 27]
```diff
@@ -542,39 +626,189 @@ def test_academy_member_id__put__passing_email(self):
...
+            model = self.bc.database.create(authenticate=True,
+                                            capability='crud_member',
+                                            role='role',
+                                            profile_academy=profile_academy)
+
+            url = reverse_lazy('authenticate:academy_member_id', kwargs={'user_id_or_email': n})
+            response = self.client.put(url)
+
+            json = response.json()
+            expected = {'detail': 'email-not-found', 'status_code': 404}
```
Comment (jefer94) [inline breathecode/authenticate/tests/urls/tests_academy_member_id.py:710]: 400

File: breathecode/authenticate/views.py [hunk 28]
```diff
@@ -271,6 +271,26 @@ def post(self, request, academy_id=None):
     @capable_of('crud_member')
     def put(self, request, academy_id=None, user_id_or_email=None):
 
+        user = ProfileAcademy.objects.filter(user__id=request.user.id, academy__id=academy_id).first()
+
+        if user.first_name is None:
+            raise ValidationException('This mentor does not have a first name',
```
Comment (jefer94) [inline breathecode/authenticate/views.py:277]: 400

File: breathecode/authenticate/views.py [hunk 29]
```diff
@@ -271,6 +271,26 @@ def post(self, request, academy_id=None):
...
+                                      code=404,
+                                      slug='first-name-not-found')
+
+        if user.last_name is None:
```
Comment (jefer94) [inline breathecode/authenticate/views.py:281]: 400

File: breathecode/authenticate/views.py [hunk 30]
```diff
@@ -271,6 +271,26 @@ def post(self, request, academy_id=None):
...
+            raise ValidationException('This mentor does not have a last name',
+                                      code=404,
+                                      slug='last-name-not-found')
+
+        if user.email is None:
```
Comment (jefer94) [inline breathecode/authenticate/views.py:286]: 400

File: breathecode/authenticate/views.py [hunk 31]
```diff
@@ -271,6 +271,26 @@ def post(self, request, academy_id=None):
...
+            raise ValidationException('This mentor does not have an email address',
+                                      code=404,
+                                      slug='email-not-found')
+
+        if user.phone is None:
+            raise ValidationException('This mentor does not have a phone', code=404, slug='phone-not-found')
```
Comment (jefer94) [inline breathecode/authenticate/views.py:292]: 400

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/authenticate/serializers.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Andresgarciasch) [reply breathecode/authenticate/serializers.py:512]: https://github.com/Andresgarciasch/apiv2/blob/ada8b0e5d16ddde3dbb5213ef03fee943561ba1a/breathecode/authenticate/tests/urls/tests_academy_member_id.py#L555

**Review by jefer94 — DISMISSED** [review]

File: breathecode/mentorship/serializers.py [hunk 32]
```diff
@@ -471,6 +473,54 @@ class Meta:
...
+
+        if 'name' not in data:
+            data['name'] = ''
+
+        if not data['name'] and profile_academy:
+
+            data['name'] = f'{profile_academy.first_name} {profile_academy.last_name}'
+
+        if not data['name']:
+            data['name'] = user.first_name + ' ' + user.last_name
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:491]: `' '` is considered as True

File: breathecode/mentorship/serializers.py [hunk 33]
```diff
@@ -471,6 +473,54 @@ class Meta:
...
+        user = data['user']
+        profile_academy = ProfileAcademy.objects.filter(user__id=data['user'].id,
+                                                        academy__id=data['academy'].id).first()
+
+        if 'name' not in data:
+            data['name'] = ''
+
+        if not data['name'] and profile_academy:
+
+            data['name'] = f'{profile_academy.first_name} {profile_academy.last_name}'
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:488]: `' '` is considered as True

File: breathecode/mentorship/serializers.py [hunk 34]
```diff
@@ -485,13 +535,50 @@ class Meta:
...
+        user = data['user'] if 'user' in data else self.instance.user
+        academy = data['academy'] if 'academy' in data else self.instance.academy
+        profile_academy = ProfileAcademy.objects.filter(user__id=user.id, academy=academy).first()
+
+        if 'name' not in data:
+            data['name'] = ''
+
+        if not data['name'] and profile_academy:
+
+            data['name'] = profile_academy.first_name + ' ' + profile_academy.last_name
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:555]: `' '` is considered as True

File: breathecode/mentorship/serializers.py [hunk 35]
```diff
@@ -485,13 +535,50 @@ class Meta:
...
+
+        if not data['name']:
+
+            data['name'] = user.first_name + ' ' + user.last_name
```
Comment (jefer94) [inline breathecode/mentorship/serializers.py:559]: `' '` is considered as True

File: breathecode/mentorship/views.py [hunk 36]
```diff
@@ -747,14 +751,36 @@ def post(self, request, academy_id=None):
...
 
         mentor = MentorProfile.objects.filter(id=mentor_id, services__academy__id=academy_id).first()
+
         if mentor is None:
             raise ValidationException('This mentor does not exist for this academy',
                                       code=404,
                                       slug='not-found')
+        user = ProfileAcademy.objects.filter(user__id=mentor.user.id, academy__id=academy_id).first()
+
+        if user.first_name is None:
```
Comment (jefer94) [inline breathecode/mentorship/views.py:767]: This doesn't work if `first_name == ''`

File: breathecode/mentorship/views.py [hunk 37]
```diff
@@ -747,14 +751,36 @@ def post(self, request, academy_id=None):
...
+            raise ValidationException(
+                translation(lang,
+                            en='This mentor does not have a first name',
+                            es='Este mentor no tiene nombre',
+                            slug='without-first-name'),
+                code=400,
+            )
+
+        if user.last_name is None:
```
Comment (jefer94) [inline breathecode/mentorship/views.py:776]: This doesn't work if `last_name == ''`

File: breathecode/authenticate/views.py [hunk 38]
```diff
@@ -270,13 +271,28 @@ def post(self, request, academy_id=None):
 
     @capable_of('crud_member')
     def put(self, request, academy_id=None, user_id_or_email=None):
+        lang = get_user_language(request)
+
+        user = ProfileAcademy.objects.filter(user__id=request.user.id, academy__id=academy_id).first()
+
+        if user.email is None:
```
Comment (jefer94) [inline breathecode/authenticate/views.py:278]: This doesn't work if `email == ''`

File: breathecode/authenticate/views.py [hunk 39]
```diff
@@ -270,13 +271,28 @@ def post(self, request, academy_id=None):
...
+            raise ValidationException(
+                translation(lang,
+                            en='This mentor does not have an email address',
+                            es='Este mentor no tiene una dirección de correo electrónico',
+                            slug='email-not-found'),
+                code=400,
+            )
+
+        if user.phone is None:
```
Comment (jefer94) [inline breathecode/authenticate/views.py:287]: This doesn't work if `phone == ''`

**Review by jefer94 — CHANGES_REQUESTED** [review]

File: breathecode/mentorship/serializers.py (hunk 34 above)
Comment (Andresgarciasch) [reply breathecode/mentorship/serializers.py:555]: there is data['name'].strip to prevent that from happenning

**Review by jefer94 — DISMISSED** [review]

**Pull request merged**